### PR TITLE
Viewport-Driven Tile Rendering

### LINKED ISSUE #19: erdblick only queries and visualizes tiles that are in Cesium viewport
For on-demand data retrieval from mapget, we need to calculate the tile IDs of the Cesium viewport. ``core/src/aabb.cpp`` implementations are there to help.

Upon initialization or a change in the viewport, erdblick should:
1. cancel any previous ongoing mapget requests, 
2. determine which tiles in the viewport are missing, and query mapget to retrieve them, 
3. determine visualised tiles which are not visible anymore and remove them.

### PR DESCRIPTION
**Pull request by josephbirkner**
Viewport-Driven Tile Rendering
==============================

This PR introduces viewport-based map data fetching for Erdblick, now making full use of the mapget `/tiles` streaming response. Closes #19. Requires [#Klebert-Engineering/mapget#](https://github.com/Klebert-Engineering/mapget/pull/30).

ezgif-3-bb4143198d

How to test (KE)?
-----------------

The viewport-driven updates are currently best tested with a GeoJSON datasource based on a directory of exported NDS.Classic tiles. Step-by-step:

1. Download the tiles from [here](https://klebertengineering-my.sharepoint.com/:u:/g/personal/j_birkner_klebert-engineering_de/EaAGW52EEjBBnq3HneGaWZEByxEwQT8JZiOgHN18QxlCqA?e=UJwuRZ). Use the  "Routing JSON Export Password" from KE-Shared to unpack the archive.
2. In Erdblick, check out this branch (`feature/viewport-driven-updates`) and run `./ci/20_linux_rebuild.bash`.
3. In your local clone of `mapget-datasource-live`, check out `release/0.2.0`.
4. With you local build of `mapget-datasource-live`, run `./livesource-server geojsonsrc --source-dir path/to/routing-json-export --with-attr-layers false`. This will launch the livesource in geojson mode on port 8889. The `--with-attr-layers false` makes the performance more akin to what can be expected with staged loading.
5. With mapviewer or mapget project, run `./mapget serve -p 8089 -d 127.0.0.1:8889 -w "path/to/erdblick/build"`.

Follow-up work
--------------

1. I added a minimal "Focus" button to map-layer entries in the maps panel. This will only work if the layer info provides a tile-id list in its `coverage` field.
2. When many data sources are connected, the frontend seems to stall occasionally when many tiles are visible. Need to change how visualisation vs parsing tasks are prioritized when many, many tiles are coming in.
3. There is no usage of WebWorkers as of yet.
6. Currently, there is a lot of performance timing console output. This should be toggle-able.
7. I have activated the Cesium OSM layer, but it should be toggle-able.
8. We should prioritize the work on staged tile loading.
9. The dot-indicators for tiles-to-be-loaded are currently not toggle-able in any way.
10. You can use wheel-click-drag to turn the camera. But I have the feeling, that the view-direction-based tile prioritization is not working. This is probably not super apparent, since the prioritization by viewport center *is* working well.

### CODE REVIEW

**Comment by josephbirkner** [comment]
Open Problems with Erdblick Viewport Updates

Tests done with NDS.Classic Munich area level 13 GeoJSON.

GeoJSON Data Source
-------------------

- [ ] Seems slow, not sure if it really reports walltime correctly.

No memory leaks observed.

Mapget Server
-------------

- [ ] There is a memory leak.
- [ ] Need to increase cache.

Erdblick Frontend
-----------------

- [ ] Update seems to start with a small delay.
      - Need predictive caching?
- [ ] Triggering update takes too much screen movement
      - Need to decrease threshold
- [ ] Update sometimes stops halfway through?
      - Need to move tile layer rendering out of AJAX callback.
      - Use WebWorker?
- [ ] Still getting "uncaught in promise" on viewport wiggle.
- [ ] Some tiles remain sometimes.
      - Need to correctly abort in-flight tiles.
- [ ] EXT_structural_metadata leads to very bad 3Dtile parsing times (10-100x)
      - Let's discuss.

No memory leaks observed.

File: static/erdblick/view.js [hunk 1]
```diff
@@ -0,0 +1,252 @@
+"use strict";
+
+import {ErdblickModel, MapViewerViewport} from "./model.js";
+import {FeatureWrapper} from "./features.js";
+
+export class ErdblickView
```
Comment (l-laura) [inline static/erdblick/view.js:6]: I remember there being some issue with the attribute panel, maybe that it disappeared unexpectedly when clicked on for navigation... is it fixed now?

File: static/erdblick/view.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        let sizeLat = north - south;
+
+        // Handle the antimeridian.
+        // TODO: Must also handle north pole.
+        if (west > -180 && sizeLon > 180.) {
+            sizeLon = 360. - sizeLon;
+        }
+
+        // Grow the viewport rectangle by 25%
+        let expandLon = sizeLon * 0.25;
```
Comment (l-laura) [inline static/erdblick/view.js:196]: Are these values used anywhere?

**Review by l-laura — CHANGES_REQUESTED** [review]
Additional notes:
1. The helpstring of mapget-datasource-live needs to be updated.
2. In the future, we should consider moving away from 1-PR-per-feature approach if they contain this many changes, in favor of PRs with smaller changes that are testable on their own.

File: libs/core/CMakeLists.txt [hunk 3]
```diff
@@ -24,6 +24,7 @@ if(${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten")
   add_executable(erdblick-core ${ERDBLICK_SOURCE_FILES})
   set_target_properties(erdblick-core PROPERTIES LINK_FLAGS "\
     --bind \
+    --profiling \
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/CMakeLists.txt:27]: Hmm, there is nothing like an actual WASM debug build. I did some tests, the erdblick-core.wasm is 1.9 MB with profiling info vs 1.2 MB without. For now, I just move the option into a comment.

File: libs/core/include/erdblick/aabb.h [hunk 4]
```diff
@@ -93,17 +92,24 @@ class Wgs84AABB
...
+    /**
+     * Returns a tile priority function based on the given camera position
+     * in WGS84, and orientation (bearing) in Radians. This priority function
+     * may be plugged into tileIdsWithPriority.
      */
-    std::vector<TileId> tileIds(
-        uint16_t level,
-        std::function<double(TileId const&)> const& tilePenaltyFun,
-        size_t limit) const;
+    static TilePriorityFn radialDistancePrioFn(glm::vec2 camPos, float orientation);
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/include/erdblick/aabb.h:112]: Good point

File: libs/core/include/erdblick/stream.h [hunk 5]
```diff
@@ -9,13 +9,51 @@ namespace erdblick
...
+     * Reset the parser by removing any buffered unparsed stream chunks.
+     */
+    void reset();
+
+    /**
+     * Access the field id dictionary offsets as currently known by this parser.
+     * This is used to tell the server whether additional field-id mapping updates
+     * need to be sent.
+     */
+    mapget::TileLayerStream::FieldOffsetMap fieldDictOffsets();
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/include/erdblick/stream.h:40]: A temporary value, by intention.

File: libs/core/src/aabb.cpp [hunk 6]
```diff
@@ -204,18 +182,17 @@ TilePriorityFn Wgs84AABB::radialDistancePrioFn(glm::vec2 camPos, float orientati
...
-        if (angle > M_PI)
-            angle = M_2_PI - angle;
+        angle -= orientation + glm::two_pi<float>();  // Difference w/ compass direction normalized from North to East
+        angle = glm::abs(glm::mod(angle, glm::two_pi<float>()));  // Map angle to circle
+        if (angle > glm::pi<float>())
+            angle = glm::two_pi<float>() - angle;
 
-        auto distance = yDiff + xDiff;  // eventually use manhattan distance to avoid comp overhead?
-        return yDiff + xDiff + angle * distance;
+        auto distance = glm::sqrt(yDiff*yDiff + xDiff*xDiff); // Use manhattan distance to avoid comp overhead?
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/src/aabb.cpp:194]: I think we can remove it. It was a TODO already in the old mapviewer code, but the sqrt here is really fast enough.

File: libs/core/src/bindings.cpp [hunk 7]
```diff
@@ -12,13 +12,94 @@
...
+ *                 - height: The height of the viewport.
+ *                 - camPosLon: The longitude of the camera position.
+ *                 - camPosLat: The latitude of the camera position.
+ *                 - orientation: The orientation of the viewport.
+ * @param level The zoom level for which to get the tile IDs.
+ * @param limit The maximum number of tile IDs to return.
+ *
+ * @return An emscripten value representing an array of prioritized tile IDs.
+ */
+em::val getTileIds(em::val viewport, int level, int limit)
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/src/bindings.cpp:44]: Agreed. I am changing the `MapViewerViewport` from model.js to an emscripten [value object type](https://emscripten.org/docs/api_reference/bind.h.html#_CPPv412value_object). For those, no explicit deletion management is needed (the drawback is that a value object is always passed by value, but that is very acceptable here I think). 😸

File: static/erdblick/view.js (hunk 1 above)
Comment (josephbirkner) [reply static/erdblick/view.js:6]: I think so 👀

File: libs/core/src/bindings.cpp [hunk 8]
```diff
@@ -12,13 +12,94 @@
...
 {
-    return Wgs84AABB::fromCenterAndTileLimit(Wgs84Point{x, y, 0}, softLimit, level);
+    double vpSouth = viewport["south"].as<double>();
+    double vpWest = viewport["west"].as<double>();
+    double vpWidth = viewport["width"].as<double>();
+    double vpHeight = viewport["height"].as<double>();
+    double camPosLon = viewport["camPosLon"].as<double>();
+    double camPosLat = viewport["camPosLat"].as<double>();
+    double orientation = viewport["orientation"].as<double>();
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/src/bindings.cpp:52]: Yup, value object type it is.

File: libs/core/src/bindings.cpp [hunk 9]
```diff
@@ -12,13 +12,94 @@
...
+    tileKey.layer_ = mapget::LayerType::Features;
+    tileKey.mapId_ = mapId;
+    tileKey.layerId_ = layerId;
+    tileKey.tileId_ = tileId;
+    return tileKey.toString();
 }
 
 EMSCRIPTEN_BINDINGS(FeatureLayerRendererBind)
 {
+    mapget::log().set_level(spdlog::level::debug);
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply libs/core/src/bindings.cpp:101]: Ok, maybe not.

File: static/erdblick/model.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        }
+        this.loadedTileLayers = newTileLayers;
+
+        // Request non-present required tile layers.
+        let requests = []
+        for (let [mapName, map] of Object.entries(this.maps)) {
+            for (let [layerName, layer] of Object.entries(map.layers))
+            {
+                // Find tile IDs which are not yet loaded for this map layer combination.
+                let requestTilesForMapLayer = []
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply static/erdblick/model.js:156]: Re-added semicolons 👍

File: static/erdblick/view.js [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        this.mouseHandler = new Cesium.ScreenSpaceEventHandler(this.viewer.scene.canvas);
+
+        // Holds the currently selected feature.
+        this.selectionTopic = new rxjs.BehaviorSubject(null); // {FeatureWrapper}
+
+        // Add a handler for selection.
+        this.mouseHandler.setInputAction(movement => {
+            // If there was a previously picked feature, reset its color.
+            if (this.pickedFeature) {
+                this.pickedFeature.color = Cesium.Color.WHITE; // Assuming the original color is WHITE. Adjust as necessary.
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply static/erdblick/view.js:52]: Very good point!

File: static/erdblick/view.js (hunk 2 above)
Comment (josephbirkner) [reply static/erdblick/view.js:196]: They were supposed to be 👍

File: static/erdblick/view.js [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+            if (feature instanceof Cesium.Cesium3DTileFeature) {
+                if (feature !== this.pickedFeature) {
+                    feature.color = Cesium.Color.GREEN;
+                    this.hoveredFeature = feature; // Store the hovered feature.
+                }
+            }
+        }, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
+
+        // Add a handler for camera movement
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply static/erdblick/view.js:84]: .

File: static/erdblick/view.js [hunk 13]
```diff
@@ -195,17 +197,20 @@ export class ErdblickView
...
         let expandLat = sizeLat * 0.25;
-        if (west > east) {
-            sizeLon += 360.;
-        }
-
-        let viewport = new MapViewerViewport(south, west, sizeLon, sizeLat, centerLon, centerLat, this.viewer.camera.heading);
-        this.model.setViewport(viewport);
+        this.model.setViewport({
+            south: south - expandLat,
+            west: west - expandLat,
```
Comment (l-laura) [inline static/erdblick/view.js:202]:
```suggestion
            west: west - expandLon,
```
And do we not need to worry about the value wrapping around somehow?

File: libs/core/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (l-laura) [reply libs/core/CMakeLists.txt]: Haven't used the output myself, but if it's useful, I'd say we can keep `--profiling` as long as it does not impact runtime and the extra size is not too tragic. At least during the initial development of erdblick.

File: libs/core/src/bindings.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (l-laura) [reply libs/core/src/bindings.cpp]: It does look much more maintainable!

File: static/erdblick/view.js (hunk 13 above)
Comment (josephbirkner) [reply static/erdblick/view.js:202]: Nope, the AABB class takes care of that.

**Comment by MisterGC** [comment]
I've tested the latest PR with various data types and sources, and it's largely stable, although I did notice occasional freezes. I believe it's in a good state for release to let users explore their own data and offer feedback.

However, my tests with the full 500-tile viewport of feature-rich data highlighted the need for staged loading. Here are some suggestions for immediate improvements in subsequent releases - that can/should be done in addition to the staged loading:

1. Performance: The GUI should remain responsive, even when handling large datasets. Lag or temporary freezes should be eliminated.
2. Loading Indicators: A more explicit visual cue is needed for tile-loading. While the current dot visualization shows the query area, it doesn't indicate if tiles are being loaded or not.
3. Performance Metrics: Display real-time statistics such as the rate of fetched features per second and the total number of features currently displayed. This will help users understand the scale of the data they're working with and facilitate more informed discussions about performance.
4. Documentation: It's crucial to document the workings of the viewport-driven update mechanism, given its complexity and involvement of multiple components and inter-process communication. This system also handles event cancellations for incoming viewport updates and requires proper request tracking. Understanding the lifetime management of various data representations—like features, attributes, and GLTF representations —is equally important. I envision a high-level overview that outlines these key behavioral aspects, serving as an entry point with references to the source code. This will equip potential contributors with a foundational understanding of the essential concepts before diving deeper into the codebase.

**Review by MisterGC — APPROVED** [review]
I've tested the latest PR with various data types and sources, and it's largely stable, although I did notice occasional freezes. I believe it's in a good state for release to let users explore their own data and offer feedback.

However, my tests with the full 500-tile viewport of feature-rich data highlighted the need for staged loading. Here are some suggestions for immediate improvements in subsequent releases - that can/should be done in addition to the staged loading:

1. **Performance**: The GUI should remain responsive, even when handling large datasets. Lag or temporary freezes should be eliminated.
2. **Loading Indicators**: A more explicit visual cue is needed for tile-loading. While the current dot visualization shows the query area, it doesn't indicate if tiles are being loaded or not.
3. **Performance Metrics**: Display real-time statistics such as the rate of fetched features per second and the total number of features currently displayed. This will help users understand the scale of the data they're working with and facilitate more informed discussions about performance.
4. **Documentation**: It's crucial to document the workings of the viewport-driven update mechanism, given its complexity and involvement of multiple components and inter-process communication. This system also handles event cancellations for incoming viewport updates and requires proper request tracking. Understanding the lifetime management of various data representations—like features, attributes, and GLTF representations —is equally important. I envision a high-level overview that outlines these key behavioral aspects, serving as an entry point with references to the source code. This will equip potential contributors with a foundational understanding of the essential concepts before diving deeper into the codebase.

**Comment by josephbirkner** [comment]
Thanks for the reviews @l-laura and @MisterGC !

**Pull request merged**